{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "2228d797-9eb5-4e5e-8f84-565deb7c17b6",
   "metadata": {},
   "source": [
    "# Greta complete Scenic+ workflow\n",
    "This notebook goes with its corresponding markdown file, with the same name.\n",
    "\n",
    "`./runi dts/hg38/pbmc10k/cases/all/runs/o_scenicplus.o_scenicplus.o_scenicplus.o_scenicplus.mdl.csv`\n",
    "\n",
    "to regenerate the DAG : `snakemake dts/hg38/pbmc10k/cases/all/runs/o_scenicplus.o_scenicplus.o_scenicplus.o_scenicplus.mdl.csv --dag | dot -Tsvg > dag.svg` "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "69e0098e-5110-412d-82d3-07cfe2e146a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import os\n",
    "import pyarrow.feather as feather\n",
    "import pyarrow as pa\n",
    "import scanpy as sc\n",
    "import muon as mu\n",
    "import scipy.io\n",
    "import h5py"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4472e98e-c550-45f1-8b08-4e681db72444",
   "metadata": {},
   "source": [
    "Here are the jobs triggered by the rules : \n",
    "```bash\n",
    "job                      count\n",
    "---------------------  -------\n",
    "annotate_pbmc10k             1\n",
    "callpeaks_pbmc10k            1\n",
    "cre_blacklist                1\n",
    "dwn_image                    2\n",
    "extract_case                 1\n",
    "gen_genome_scenicplus        1\n",
    "gen_gid_ensmbl               1\n",
    "gen_motif_scenicplus         1\n",
    "mdl_o_scenicplus             1\n",
    "prcannot_pbmc10k             1\n",
    "total                       11\n",
    "``` "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a45e3ede-a2d3-4893-aa17-ef6c431c351d",
   "metadata": {},
   "source": [
    "## 1. Inputs "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e162bb4a-4078-4043-9038-cc94bbf5ff8d",
   "metadata": {},
   "outputs": [],
   "source": [
    "dir = \"/globalscratch/vangysel/greta\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "cb07f425-4074-40c2-bb38-2a9c0b19dee8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" width=\"527pt\" height=\"420pt\" viewBox=\"0.00 0.00 527.41 420.00\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(1 1) rotate(0) translate(4 416)\">\n",
       "<title>snakemake_dag</title>\n",
       "<polygon fill=\"#ffffff\" stroke=\"transparent\" points=\"-4,4 -4,-416 523.4128,-416 523.4128,4 -4,4\"/>\n",
       "<!-- 0 -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>0</title>\n",
       "<path fill=\"none\" stroke=\"#80d856\" stroke-width=\"2\" d=\"M265.9128,-36C265.9128,-36 186.9128,-36 186.9128,-36 180.9128,-36 174.9128,-30 174.9128,-24 174.9128,-24 174.9128,-12 174.9128,-12 174.9128,-6 180.9128,0 186.9128,0 186.9128,0 265.9128,0 265.9128,0 271.9128,0 277.9128,-6 277.9128,-12 277.9128,-12 277.9128,-24 277.9128,-24 277.9128,-30 271.9128,-36 265.9128,-36\"/>\n",
       "<text text-anchor=\"middle\" x=\"226.4128\" y=\"-15.5\" font-family=\"sans\" font-size=\"10.00\" fill=\"#000000\">mdl_o_scenicplus</text>\n",
       "</g>\n",
       "<!-- 1 -->\n",
       "<g id=\"node2\" class=\"node\">\n",
       "<title>1</title>\n",
       "<path fill=\"none\" stroke=\"#5663d8\" stroke-width=\"2\" stroke-dasharray=\"5,2\" d=\"M129.9128,-196C129.9128,-196 28.9128,-196 28.9128,-196 22.9128,-196 16.9128,-190 16.9128,-184 16.9128,-184 16.9128,-172 16.9128,-172 16.9128,-166 22.9128,-160 28.9128,-160 28.9128,-160 129.9128,-160 129.9128,-160 135.9128,-160 141.9128,-166 141.9128,-172 141.9128,-172 141.9128,-184 141.9128,-184 141.9128,-190 135.9128,-196 129.9128,-196\"/>\n",
       "<text text-anchor=\"middle\" x=\"79.4128\" y=\"-181\" font-family=\"sans\" font-size=\"10.00\" fill=\"#000000\">dwn_image</text>\n",
       "<text text-anchor=\"middle\" x=\"79.4128\" y=\"-170\" font-family=\"sans\" font-size=\"10.00\" fill=\"#000000\">name_img: scenicplus</text>\n",
       "</g>\n",
       "<!-- 1&#45;&gt;0 -->\n",
       "<g id=\"edge1\" class=\"edge\">\n",
       "<title>1-&gt;0</title>\n",
       "<path fill=\"none\" stroke=\"#c0c0c0\" stroke-width=\"2\" d=\"M43.9774,-159.8842C30.5787,-151.0325 16.7929,-139.0142 9.4128,-124 -.7823,-103.2591 -4.82,-90.2086 9.4128,-72 28.4148,-47.6899 107.4823,-32.7498 164.6973,-24.9162\"/>\n",
       "<polygon fill=\"#c0c0c0\" stroke=\"#c0c0c0\" stroke-width=\"2\" points=\"165.2968,-28.3674 174.749,-23.5815 164.3753,-21.4283 165.2968,-28.3674\"/>\n",
       "</g>\n",
       "<!-- 11 -->\n",
       "<g id=\"node12\" class=\"node\">\n",
       "<title>11</title>\n",
       "<path fill=\"none\" stroke=\"#d87d56\" stroke-width=\"2\" stroke-dasharray=\"5,2\" d=\"M128.4128,-116C128.4128,-116 30.4128,-116 30.4128,-116 24.4128,-116 18.4128,-110 18.4128,-104 18.4128,-104 18.4128,-92 18.4128,-92 18.4128,-86 24.4128,-80 30.4128,-80 30.4128,-80 128.4128,-80 128.4128,-80 134.4128,-80 140.4128,-86 140.4128,-92 140.4128,-92 140.4128,-104 140.4128,-104 140.4128,-110 134.4128,-116 128.4128,-116\"/>\n",
       "<text text-anchor=\"middle\" x=\"79.4128\" y=\"-95.5\" font-family=\"sans\" font-size=\"10.00\" fill=\"#000000\">gen_motif_scenicplus</text>\n",
       "</g>\n",
       "<!-- 1&#45;&gt;11 -->\n",
       "<g id=\"edge16\" class=\"edge\">\n",
       "<title>1-&gt;11</title>\n",
       "<path fill=\"none\" stroke=\"#c0c0c0\" stroke-width=\"2\" d=\"M79.4128,-159.8468C79.4128,-149.9251 79.4128,-137.4167 79.4128,-126.2492\"/>\n",
       "<polygon fill=\"#c0c0c0\" stroke=\"#c0c0c0\" stroke-width=\"2\" points=\"82.9129,-126.0814 79.4128,-116.0814 75.9129,-126.0814 82.9129,-126.0814\"/>\n",
       "</g>\n",
       "<!-- 12 -->\n",
       "<g id=\"node13\" class=\"node\">\n",
       "<title>12</title>\n",
       "<path fill=\"none\" stroke=\"#d6d856\" stroke-width=\"2\" stroke-dasharray=\"5,2\" d=\"M282.4128,-116C282.4128,-116 170.4128,-116 170.4128,-116 164.4128,-116 158.4128,-110 158.4128,-104 158.4128,-104 158.4128,-92 158.4128,-92 158.4128,-86 164.4128,-80 170.4128,-80 170.4128,-80 282.4128,-80 282.4128,-80 288.4128,-80 294.4128,-86 294.4128,-92 294.4128,-92 294.4128,-104 294.4128,-104 294.4128,-110 288.4128,-116 282.4128,-116\"/>\n",
       "<text text-anchor=\"middle\" x=\"226.4128\" y=\"-95.5\" font-family=\"sans\" font-size=\"10.00\" fill=\"#000000\">gen_genome_scenicplus</text>\n",
       "</g>\n",
       "<!-- 1&#45;&gt;12 -->\n",
       "<g id=\"edge17\" class=\"edge\">\n",
       "<title>1-&gt;12</title>\n",
       "<path fill=\"none\" stroke=\"#c0c0c0\" stroke-width=\"2\" d=\"M112.7694,-159.8468C133.9136,-148.3397 161.4515,-133.3531 184.1236,-121.0145\"/>\n",
       "<polygon fill=\"#c0c0c0\" stroke=\"#c0c0c0\" stroke-width=\"2\" points=\"186.0778,-123.9358 193.1882,-116.0814 182.7316,-117.7874 186.0778,-123.9358\"/>\n",
       "</g>\n",
       "<!-- 2 -->\n",
       "<g id=\"node3\" class=\"node\">\n",
       "<title>2</title>\n",
       "<path fill=\"none\" stroke=\"#56d8b9\" stroke-width=\"2\" d=\"M383.9128,-124C383.9128,-124 324.9128,-124 324.9128,-124 318.9128,-124 312.9128,-118 312.9128,-112 312.9128,-112 312.9128,-84 312.9128,-84 312.9128,-78 318.9128,-72 324.9128,-72 324.9128,-72 383.9128,-72 383.9128,-72 389.9128,-72 395.9128,-78 395.9128,-84 395.9128,-84 395.9128,-112 395.9128,-112 395.9128,-118 389.9128,-124 383.9128,-124\"/>\n",
       "<text text-anchor=\"middle\" x=\"354.4128\" y=\"-112\" font-family=\"sans\" font-size=\"10.00\" fill=\"#000000\">extract_case</text>\n",
       "<text text-anchor=\"middle\" x=\"354.4128\" y=\"-101\" font-family=\"sans\" font-size=\"10.00\" fill=\"#000000\">case: all</text>\n",
       "<text text-anchor=\"middle\" x=\"354.4128\" y=\"-90\" font-family=\"sans\" font-size=\"10.00\" fill=\"#000000\">dat: pbmc10k</text>\n",
       "<text text-anchor=\"middle\" x=\"354.4128\" y=\"-79\" font-family=\"sans\" font-size=\"10.00\" fill=\"#000000\">org: hg38</text>\n",
       "</g>\n",
       "<!-- 2&#45;&gt;0 -->\n",
       "<g id=\"edge2\" class=\"edge\">\n",
       "<title>2-&gt;0</title>\n",
       "<path fill=\"none\" stroke=\"#c0c0c0\" stroke-width=\"2\" d=\"M312.8549,-72.0263C297.1112,-62.1865 279.2929,-51.0501 263.9414,-41.4554\"/>\n",
       "<polygon fill=\"#c0c0c0\" stroke=\"#c0c0c0\" stroke-width=\"2\" points=\"265.557,-38.3378 255.222,-36.0058 261.847,-44.2738 265.557,-38.3378\"/>\n",
       "</g>\n",
       "<!-- 3 -->\n",
       "<g id=\"node4\" class=\"node\">\n",
       "<title>3</title>\n",
       "<path fill=\"none\" stroke=\"#567bd8\" stroke-width=\"2\" d=\"M366.4128,-196C366.4128,-196 280.4128,-196 280.4128,-196 274.4128,-196 268.4128,-190 268.4128,-184 268.4128,-184 268.4128,-172 268.4128,-172 268.4128,-166 274.4128,-160 280.4128,-160 280.4128,-160 366.4128,-160 366.4128,-160 372.4128,-160 378.4128,-166 378.4128,-172 378.4128,-172 378.4128,-184 378.4128,-184 378.4128,-190 372.4128,-196 366.4128,-196\"/>\n",
       "<text text-anchor=\"middle\" x=\"323.4128\" y=\"-175.5\" font-family=\"sans\" font-size=\"10.00\" fill=\"#000000\">annotate_pbmc10k</text>\n",
       "</g>\n",
       "<!-- 3&#45;&gt;2 -->\n",
       "<g id=\"edge6\" class=\"edge\">\n",
       "<title>3-&gt;2</title>\n",
       "<path fill=\"none\" stroke=\"#c0c0c0\" stroke-width=\"2\" d=\"M330.4472,-159.8468C333.4382,-152.1278 337.036,-142.8432 340.5238,-133.8425\"/>\n",
       "<polygon fill=\"#c0c0c0\" stroke=\"#c0c0c0\" stroke-width=\"2\" points=\"343.8816,-134.8637 344.2314,-124.2746 337.3545,-132.3344 343.8816,-134.8637\"/>\n",
       "</g>\n",
       "<!-- 4 -->\n",
       "<g id=\"node5\" class=\"node\">\n",
       "<title>4</title>\n",
       "<path fill=\"none\" stroke=\"#56c1d8\" stroke-width=\"2\" stroke-dasharray=\"5,2\" d=\"M342.4128,-340C342.4128,-340 256.4128,-340 256.4128,-340 250.4128,-340 244.4128,-334 244.4128,-328 244.4128,-328 244.4128,-316 244.4128,-316 244.4128,-310 250.4128,-304 256.4128,-304 256.4128,-304 342.4128,-304 342.4128,-304 348.4128,-304 354.4128,-310 354.4128,-316 354.4128,-316 354.4128,-328 354.4128,-328 354.4128,-334 348.4128,-340 342.4128,-340\"/>\n",
       "<text text-anchor=\"middle\" x=\"299.4128\" y=\"-319.5\" font-family=\"sans\" font-size=\"10.00\" fill=\"#000000\">prcannot_pbmc10k</text>\n",
       "</g>\n",
       "<!-- 4&#45;&gt;3 -->\n",
       "<g id=\"edge7\" class=\"edge\">\n",
       "<title>4-&gt;3</title>\n",
       "<path fill=\"none\" stroke=\"#c0c0c0\" stroke-width=\"2\" d=\"M302.4524,-303.7623C306.564,-279.0928 313.9362,-234.8598 318.7506,-205.9731\"/>\n",
       "<polygon fill=\"#c0c0c0\" stroke=\"#c0c0c0\" stroke-width=\"2\" points=\"322.2062,-206.529 320.3978,-196.0896 315.3014,-205.3781 322.2062,-206.529\"/>\n",
       "</g>\n",
       "<!-- 6 -->\n",
       "<g id=\"node7\" class=\"node\">\n",
       "<title>6</title>\n",
       "<path fill=\"none\" stroke=\"#9fd856\" stroke-width=\"2\" d=\"M258.9128,-268C258.9128,-268 169.9128,-268 169.9128,-268 163.9128,-268 157.9128,-262 157.9128,-256 157.9128,-256 157.9128,-244 157.9128,-244 157.9128,-238 163.9128,-232 169.9128,-232 169.9128,-232 258.9128,-232 258.9128,-232 264.9128,-232 270.9128,-238 270.9128,-244 270.9128,-244 270.9128,-256 270.9128,-256 270.9128,-262 264.9128,-268 258.9128,-268\"/>\n",
       "<text text-anchor=\"middle\" x=\"214.4128\" y=\"-247.5\" font-family=\"sans\" font-size=\"10.00\" fill=\"#000000\">callpeaks_pbmc10k</text>\n",
       "</g>\n",
       "<!-- 4&#45;&gt;6 -->\n",
       "<g id=\"edge12\" class=\"edge\">\n",
       "<title>4-&gt;6</title>\n",
       "<path fill=\"none\" stroke=\"#c0c0c0\" stroke-width=\"2\" d=\"M277.9637,-303.8314C267.595,-295.0485 254.9899,-284.3712 243.7482,-274.8489\"/>\n",
       "<polygon fill=\"#c0c0c0\" stroke=\"#c0c0c0\" stroke-width=\"2\" points=\"245.7477,-271.9556 235.855,-268.1628 241.2233,-277.297 245.7477,-271.9556\"/>\n",
       "</g>\n",
       "<!-- 5 -->\n",
       "<g id=\"node6\" class=\"node\">\n",
       "<title>5</title>\n",
       "<path fill=\"none\" stroke=\"#5663d8\" stroke-width=\"2\" stroke-dasharray=\"5,2\" d=\"M434.9128,-412C434.9128,-412 329.9128,-412 329.9128,-412 323.9128,-412 317.9128,-406 317.9128,-400 317.9128,-400 317.9128,-388 317.9128,-388 317.9128,-382 323.9128,-376 329.9128,-376 329.9128,-376 434.9128,-376 434.9128,-376 440.9128,-376 446.9128,-382 446.9128,-388 446.9128,-388 446.9128,-400 446.9128,-400 446.9128,-406 440.9128,-412 434.9128,-412\"/>\n",
       "<text text-anchor=\"middle\" x=\"382.4128\" y=\"-397\" font-family=\"sans\" font-size=\"10.00\" fill=\"#000000\">dwn_image</text>\n",
       "<text text-anchor=\"middle\" x=\"382.4128\" y=\"-386\" font-family=\"sans\" font-size=\"10.00\" fill=\"#000000\">name_img: gretabench</text>\n",
       "</g>\n",
       "<!-- 5&#45;&gt;4 -->\n",
       "<g id=\"edge10\" class=\"edge\">\n",
       "<title>5-&gt;4</title>\n",
       "<path fill=\"none\" stroke=\"#c0c0c0\" stroke-width=\"2\" d=\"M361.4684,-375.8314C351.3437,-367.0485 339.0351,-356.3712 328.058,-346.8489\"/>\n",
       "<polygon fill=\"#c0c0c0\" stroke=\"#c0c0c0\" stroke-width=\"2\" points=\"330.1979,-344.0718 320.3505,-340.1628 325.6109,-349.3595 330.1979,-344.0718\"/>\n",
       "</g>\n",
       "<!-- 9 -->\n",
       "<g id=\"node10\" class=\"node\">\n",
       "<title>9</title>\n",
       "<path fill=\"none\" stroke=\"#56d863\" stroke-width=\"2\" stroke-dasharray=\"5,2\" d=\"M411.4128,-268C411.4128,-268 339.4128,-268 339.4128,-268 333.4128,-268 327.4128,-262 327.4128,-256 327.4128,-256 327.4128,-244 327.4128,-244 327.4128,-238 333.4128,-232 339.4128,-232 339.4128,-232 411.4128,-232 411.4128,-232 417.4128,-232 423.4128,-238 423.4128,-244 423.4128,-244 423.4128,-256 423.4128,-256 423.4128,-262 417.4128,-268 411.4128,-268\"/>\n",
       "<text text-anchor=\"middle\" x=\"375.4128\" y=\"-247.5\" font-family=\"sans\" font-size=\"10.00\" fill=\"#000000\">gen_gid_ensmbl</text>\n",
       "</g>\n",
       "<!-- 5&#45;&gt;9 -->\n",
       "<g id=\"edge14\" class=\"edge\">\n",
       "<title>5-&gt;9</title>\n",
       "<path fill=\"none\" stroke=\"#c0c0c0\" stroke-width=\"2\" d=\"M381.5262,-375.7623C380.3323,-351.201 378.1956,-307.2474 376.7911,-278.3541\"/>\n",
       "<polygon fill=\"#c0c0c0\" stroke=\"#c0c0c0\" stroke-width=\"2\" points=\"380.2736,-277.9079 376.2921,-268.0896 373.2819,-278.2478 380.2736,-277.9079\"/>\n",
       "</g>\n",
       "<!-- 10 -->\n",
       "<g id=\"node11\" class=\"node\">\n",
       "<title>10</title>\n",
       "<path fill=\"none\" stroke=\"#d6d856\" stroke-width=\"2\" stroke-dasharray=\"5,2\" d=\"M507.4128,-268C507.4128,-268 453.4128,-268 453.4128,-268 447.4128,-268 441.4128,-262 441.4128,-256 441.4128,-256 441.4128,-244 441.4128,-244 441.4128,-238 447.4128,-232 453.4128,-232 453.4128,-232 507.4128,-232 507.4128,-232 513.4128,-232 519.4128,-238 519.4128,-244 519.4128,-244 519.4128,-256 519.4128,-256 519.4128,-262 513.4128,-268 507.4128,-268\"/>\n",
       "<text text-anchor=\"middle\" x=\"480.4128\" y=\"-247.5\" font-family=\"sans\" font-size=\"10.00\" fill=\"#000000\">cre_blacklist</text>\n",
       "</g>\n",
       "<!-- 5&#45;&gt;10 -->\n",
       "<g id=\"edge15\" class=\"edge\">\n",
       "<title>5-&gt;10</title>\n",
       "<path fill=\"none\" stroke=\"#c0c0c0\" stroke-width=\"2\" d=\"M394.8246,-375.7623C411.908,-350.66 442.7769,-305.3017 462.4005,-276.4671\"/>\n",
       "<polygon fill=\"#c0c0c0\" stroke=\"#c0c0c0\" stroke-width=\"2\" points=\"465.369,-278.326 468.1018,-268.0896 459.582,-274.3876 465.369,-278.326\"/>\n",
       "</g>\n",
       "<!-- 6&#45;&gt;3 -->\n",
       "<g id=\"edge8\" class=\"edge\">\n",
       "<title>6-&gt;3</title>\n",
       "<path fill=\"none\" stroke=\"#c0c0c0\" stroke-width=\"2\" d=\"M241.9181,-231.8314C255.7308,-222.7074 272.6384,-211.539 287.4657,-201.7449\"/>\n",
       "<polygon fill=\"#c0c0c0\" stroke=\"#c0c0c0\" stroke-width=\"2\" points=\"289.5014,-204.5949 295.9163,-196.1628 285.6432,-198.7541 289.5014,-204.5949\"/>\n",
       "</g>\n",
       "<!-- 7 -->\n",
       "<g id=\"node8\" class=\"node\">\n",
       "<title>7</title>\n",
       "<path fill=\"none\" stroke=\"#d85e56\" stroke-width=\"2\" stroke-dasharray=\"5,2\" d=\"M214.4128,-340C214.4128,-340 124.4128,-340 124.4128,-340 118.4128,-340 112.4128,-334 112.4128,-328 112.4128,-328 112.4128,-316 112.4128,-316 112.4128,-310 118.4128,-304 124.4128,-304 124.4128,-304 214.4128,-304 214.4128,-304 220.4128,-304 226.4128,-310 226.4128,-316 226.4128,-316 226.4128,-328 226.4128,-328 226.4128,-334 220.4128,-340 214.4128,-340\"/>\n",
       "<text text-anchor=\"middle\" x=\"169.4128\" y=\"-319.5\" font-family=\"sans\" font-size=\"10.00\" fill=\"#000000\">download_pbmc10k</text>\n",
       "</g>\n",
       "<!-- 7&#45;&gt;6 -->\n",
       "<g id=\"edge11\" class=\"edge\">\n",
       "<title>7-&gt;6</title>\n",
       "<path fill=\"none\" stroke=\"#c0c0c0\" stroke-width=\"2\" d=\"M180.7682,-303.8314C185.8454,-295.7079 191.9355,-285.9637 197.5304,-277.0118\"/>\n",
       "<polygon fill=\"#c0c0c0\" stroke=\"#c0c0c0\" stroke-width=\"2\" points=\"200.5725,-278.7483 202.9045,-268.4133 194.6365,-275.0383 200.5725,-278.7483\"/>\n",
       "</g>\n",
       "<!-- 8 -->\n",
       "<g id=\"node9\" class=\"node\">\n",
       "<title>8</title>\n",
       "<path fill=\"none\" stroke=\"#5663d8\" stroke-width=\"2\" stroke-dasharray=\"5,2\" d=\"M202.4128,-412C202.4128,-412 136.4128,-412 136.4128,-412 130.4128,-412 124.4128,-406 124.4128,-400 124.4128,-400 124.4128,-388 124.4128,-388 124.4128,-382 130.4128,-376 136.4128,-376 136.4128,-376 202.4128,-376 202.4128,-376 208.4128,-376 214.4128,-382 214.4128,-388 214.4128,-388 214.4128,-400 214.4128,-400 214.4128,-406 208.4128,-412 202.4128,-412\"/>\n",
       "<text text-anchor=\"middle\" x=\"169.4128\" y=\"-397\" font-family=\"sans\" font-size=\"10.00\" fill=\"#000000\">dwn_image</text>\n",
       "<text text-anchor=\"middle\" x=\"169.4128\" y=\"-386\" font-family=\"sans\" font-size=\"10.00\" fill=\"#000000\">name_img: figr</text>\n",
       "</g>\n",
       "<!-- 8&#45;&gt;7 -->\n",
       "<g id=\"edge13\" class=\"edge\">\n",
       "<title>8-&gt;7</title>\n",
       "<path fill=\"none\" stroke=\"#c0c0c0\" stroke-width=\"2\" d=\"M169.4128,-375.8314C169.4128,-368.131 169.4128,-358.9743 169.4128,-350.4166\"/>\n",
       "<polygon fill=\"#c0c0c0\" stroke=\"#c0c0c0\" stroke-width=\"2\" points=\"172.9129,-350.4132 169.4128,-340.4133 165.9129,-350.4133 172.9129,-350.4132\"/>\n",
       "</g>\n",
       "<!-- 9&#45;&gt;3 -->\n",
       "<g id=\"edge9\" class=\"edge\">\n",
       "<title>9-&gt;3</title>\n",
       "<path fill=\"none\" stroke=\"#c0c0c0\" stroke-width=\"2\" d=\"M362.291,-231.8314C356.3629,-223.6232 349.2398,-213.7606 342.7194,-204.7323\"/>\n",
       "<polygon fill=\"#c0c0c0\" stroke=\"#c0c0c0\" stroke-width=\"2\" points=\"345.4036,-202.4708 336.7113,-196.4133 339.7288,-206.5693 345.4036,-202.4708\"/>\n",
       "</g>\n",
       "<!-- 10&#45;&gt;0 -->\n",
       "<g id=\"edge3\" class=\"edge\">\n",
       "<title>10-&gt;0</title>\n",
       "<path fill=\"none\" stroke=\"#c0c0c0\" stroke-width=\"2\" d=\"M477.8339,-231.9473C471.7591,-196.123 453.1996,-115.7396 404.4128,-72 386.9765,-56.3676 332.1667,-41.116 287.9258,-30.8311\"/>\n",
       "<polygon fill=\"#c0c0c0\" stroke=\"#c0c0c0\" stroke-width=\"2\" points=\"288.6184,-27.3992 278.0901,-28.5848 287.0598,-34.2235 288.6184,-27.3992\"/>\n",
       "</g>\n",
       "<!-- 11&#45;&gt;0 -->\n",
       "<g id=\"edge4\" class=\"edge\">\n",
       "<title>11-&gt;0</title>\n",
       "<path fill=\"none\" stroke=\"#c0c0c0\" stroke-width=\"2\" d=\"M112.7694,-79.8468C133.9136,-68.3397 161.4515,-53.3531 184.1236,-41.0145\"/>\n",
       "<polygon fill=\"#c0c0c0\" stroke=\"#c0c0c0\" stroke-width=\"2\" points=\"186.0778,-43.9358 193.1882,-36.0814 182.7316,-37.7874 186.0778,-43.9358\"/>\n",
       "</g>\n",
       "<!-- 12&#45;&gt;0 -->\n",
       "<g id=\"edge5\" class=\"edge\">\n",
       "<title>12-&gt;0</title>\n",
       "<path fill=\"none\" stroke=\"#c0c0c0\" stroke-width=\"2\" d=\"M226.4128,-79.8468C226.4128,-69.9251 226.4128,-57.4167 226.4128,-46.2492\"/>\n",
       "<polygon fill=\"#c0c0c0\" stroke=\"#c0c0c0\" stroke-width=\"2\" points=\"229.9129,-46.0814 226.4128,-36.0814 222.9129,-46.0814 229.9129,-46.0814\"/>\n",
       "</g>\n",
       "</g>\n",
       "</svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from IPython.display import SVG, display\n",
    "display(SVG(filename=f\"{dir}/notes_scenic+/dag.svg\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de9e5834-9f2e-45b5-b719-cdedabc34939",
   "metadata": {},
   "source": [
    "We will go through each layer of the DAG (top layer is the `dwn_image` and last layer is `mdl_o_scenicplus`)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7d965ae6-a8a7-4aa7-a1b0-180bbcde0cd7",
   "metadata": {},
   "source": [
    "### Layer 1 \n",
    "Download `figr.sif` & `gretabench.sif` the images used to run the further steps.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c64ec1c6-6fe4-493a-acdb-25c2c46607b9",
   "metadata": {},
   "source": [
    "### Layer 2 \n",
    "\n",
    "-  `download_pbmc10k` : download the pbmc10k dataset (get ATAC files, RNA are downloaded sperately and cached)\n",
    "    - `dts/hg38/pbmc10k/smpl.frags.tsv.gz`\n",
    "    - `dts/hg38/pbmc10k/smpl.frags.tsv.gz.tbi`\n",
    "<br><br>\n",
    "      \n",
    "- `prcannot_pbmc10k` : cell annotation (metadata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "4753eca2-007a-405e-86d3-125ff9894afa",
   "metadata": {},
   "outputs": [],
   "source": [
    "frags = pd.read_csv(\n",
    "    f\"{dir}/dts/hg38/pbmc10k/smpl.frags.tsv.gz\",\n",
    "    sep=\"\\t\",\n",
    "    nrows=100000,   # do not load the entire file\n",
    "    header=None,\n",
    "    names=[\"chr\", \"start\", \"end\", \"barcode\", \"count\"]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "5e55cb72-936b-408d-87ca-b7c9a42a8cbc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>barcode</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10066</td>\n",
       "      <td>10478</td>\n",
       "      <td>smpl_TCAAGAACAGTAATAG</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10072</td>\n",
       "      <td>10191</td>\n",
       "      <td>smpl_AACCCGCAGGTAGCTT</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10073</td>\n",
       "      <td>10340</td>\n",
       "      <td>smpl_CGCATATAGGTTACGT</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr  start    end                barcode  count\n",
       "0  chr1  10066  10478  smpl_TCAAGAACAGTAATAG      1\n",
       "1  chr1  10072  10191  smpl_AACCCGCAGGTAGCTT      1\n",
       "2  chr1  10073  10340  smpl_CGCATATAGGTTACGT      2"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "frags.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "d0c3ee28-6b2e-4e22-a84f-8f01eef92129",
   "metadata": {},
   "outputs": [],
   "source": [
    "pbmc_annot = pd.read_csv(f\"{dir}/dts/hg38/pbmc10k/annot.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "64041a07-12d4-47ff-9d34-1cec9a14fb54",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>batch</th>\n",
       "      <th>celltype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>smpl_AAACAGCCAATCCCTT</td>\n",
       "      <td>smpl</td>\n",
       "      <td>CD4 TCM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>smpl_AAACAGCCAATGCGCT</td>\n",
       "      <td>smpl</td>\n",
       "      <td>CD4 Naive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>smpl_AAACAGCCACCAACCG</td>\n",
       "      <td>smpl</td>\n",
       "      <td>CD8 Naive</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              Unnamed: 0 batch   celltype\n",
       "0  smpl_AAACAGCCAATCCCTT  smpl    CD4 TCM\n",
       "1  smpl_AAACAGCCAATGCGCT  smpl  CD4 Naive\n",
       "2  smpl_AAACAGCCACCAACCG  smpl  CD8 Naive"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pbmc_annot.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "377edd11-0ad8-4dfb-8bf5-0bd8c88aa93d",
   "metadata": {},
   "source": [
    "### Layer 3\n",
    "- `gen_gid_ensmbl` : gene id symbol mapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "9ba5776c-215d-4016-866b-0a98db044081",
   "metadata": {},
   "outputs": [],
   "source": [
    "ens_human = pd.read_csv(f\"{dir}/dbs/hg38/gen/gid/ensembl.csv\")\n",
    "ens_mouse = pd.read_csv(f\"{dir}/dbs/mm10/gen/gid/ensembl.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "4cc50663-88f1-4254-9fba-1908aa3234ab",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>symbol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000210049</td>\n",
       "      <td>MT-TF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000211459</td>\n",
       "      <td>MT-RNR1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000210077</td>\n",
       "      <td>MT-TV</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                id   symbol\n",
       "0  ENSG00000210049    MT-TF\n",
       "1  ENSG00000211459  MT-RNR1\n",
       "2  ENSG00000210077    MT-TV"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ens_human.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "17d1e66e-54c4-4657-b61b-867c030c218d",
   "metadata": {},
   "source": [
    "- `cre_blacklist` : regions to ignore"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "c59f6ce8-31e3-49df-9292-981683c082a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "blacklist = pd.read_csv(\n",
    "    f\"{dir}/dbs/hg38/cre/blacklist/blacklist.bed.gz\",\n",
    "    sep=\"\\t\",\n",
    "    header=None,\n",
    "    names=[\"start\", \"end\", \"reason\"]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "8a1a35d9-6654-4ef8-8aad-38bcd73d510d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>reason</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr10</th>\n",
       "      <td>0</td>\n",
       "      <td>45700</td>\n",
       "      <td>Low Mappability</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr10</th>\n",
       "      <td>38481300</td>\n",
       "      <td>38596500</td>\n",
       "      <td>High Signal Region</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr10</th>\n",
       "      <td>38782600</td>\n",
       "      <td>38967900</td>\n",
       "      <td>High Signal Region</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          start       end              reason\n",
       "chr10         0     45700     Low Mappability\n",
       "chr10  38481300  38596500  High Signal Region\n",
       "chr10  38782600  38967900  High Signal Region"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "blacklist.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "12760117-08b1-477b-8722-4caf31160b12",
   "metadata": {},
   "source": [
    "- `callpeaks_pbmc10k`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "bc0c9a48-20a6-4a1d-8b53-50f87910baab",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 9631 × 215676"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "peaks = sc.read_h5ad(f\"{dir}/dts/hg38/pbmc10k/peaks.h5ad\")\n",
    "peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "810e2caa-ab4b-4c13-bd12-02092b27bfa6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>smpl_AAACAGCCAATCCCTT</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>smpl_AAACAGCCAATGCGCT</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>smpl_AAACAGCCACCAACCG</th>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: [smpl_AAACAGCCAATCCCTT, smpl_AAACAGCCAATGCGCT, smpl_AAACAGCCACCAACCG]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "peaks.obs.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "bd5791e1-01dc-447c-87b1-2b3e9d7ed006",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr1-9947-10447</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1-96340-96840</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1-180660-181160</th>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: [chr1-9947-10447, chr1-96340-96840, chr1-180660-181160]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "peaks.var.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1fc0e85a-9b82-43a0-87ae-b53af75b72d6",
   "metadata": {},
   "source": [
    "### Layer 4 \n",
    "- `annotate_pbmc10k`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "292be008-e91a-4c3b-b9f7-62b3f8ad64fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# assembles the raw multimodal object <annotated.h5mu>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "d5728309-eddd-4444-b42d-765b19017654",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ucl/inma/vangysel/.local/lib/python3.11/site-packages/mudata/_core/mudata.py:1403: FutureWarning: From 0.4 .update() will not pull obs/var columns from individual modalities by default anymore. Set mudata.set_options(pull_on_update=False) to adopt the new behaviour, which will become the default. Use new pull_obs/pull_var and push_obs/push_var methods for more flexibility.\n",
      "  self._update_attr(\"var\", axis=0, join_common=join_common)\n",
      "/home/ucl/inma/vangysel/.local/lib/python3.11/site-packages/mudata/_core/mudata.py:1275: FutureWarning: From 0.4 .update() will not pull obs/var columns from individual modalities by default anymore. Set mudata.set_options(pull_on_update=False) to adopt the new behaviour, which will become the default. Use new pull_obs/pull_var and push_obs/push_var methods for more flexibility.\n",
      "  self._update_attr(\"obs\", axis=1, join_common=join_common)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<pre>MuData object with n_obs × n_vars = 9631 × 233968\n",
       "  obs:\t&#x27;batch&#x27;, &#x27;celltype&#x27;\n",
       "  2 modalities\n",
       "    rna:\t9631 x 18292\n",
       "    atac:\t9631 x 215676</pre>"
      ],
      "text/plain": [
       "MuData object with n_obs × n_vars = 9631 × 233968\n",
       "  obs:\t'batch', 'celltype'\n",
       "  2 modalities\n",
       "    rna:\t9631 x 18292\n",
       "    atac:\t9631 x 215676"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annotated = mu.read(f\"{dir}/dts/hg38/pbmc10k/annotated.h5mu\")\n",
    "annotated"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "1ff4e634-1f1f-4fa5-9168-c9992ffa74c9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 9631 × 18292"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annotated['rna']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "519f4251-4895-480a-8424-940f873f8d28",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>smpl_AAACAGCCAATCCCTT</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>smpl_AAACAGCCAATGCGCT</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>smpl_AAACAGCCACCAACCG</th>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: [smpl_AAACAGCCAATCCCTT, smpl_AAACAGCCAATGCGCT, smpl_AAACAGCCACCAACCG]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annotated['rna'].obs.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "1bdff170-4792-4b30-8472-39a08e37c720",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>LINC01409</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FAM87B</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LINC01128</th>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: [LINC01409, FAM87B, LINC01128]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annotated['rna'].var.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "b3dfd743-5e70-4895-a625-6af45c889dd0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 9631 × 215676"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annotated['atac']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "59695249-098b-4518-9c82-6d631dfcd301",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>smpl_AAACAGCCAATCCCTT</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>smpl_AAACAGCCAATGCGCT</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>smpl_AAACAGCCACCAACCG</th>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: [smpl_AAACAGCCAATCCCTT, smpl_AAACAGCCAATGCGCT, smpl_AAACAGCCACCAACCG]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annotated['atac'].obs.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "d2984faa-d672-4c14-904e-bc0fa609928c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr1-9947-10447</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1-96340-96840</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1-180660-181160</th>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: [chr1-9947-10447, chr1-96340-96840, chr1-180660-181160]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annotated['atac'].var.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad0c370b-69d0-486c-8a29-1aa538fe93b8",
   "metadata": {},
   "source": [
    "### Layer 5\n",
    "\n",
    "The PBMC10k dataset is now ready (`.h5mu` file), we can run `mdl_o_scenicplus`.  <br><br>\n",
    "It needs the following inputs : <br>\n",
    "2) `blist` : cre blacklist (already download in Layer 3 : `cre_blacklist`)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a6864506-26e3-49eb-8dbf-c79c24305b68",
   "metadata": {},
   "source": [
    "#### 5.1 From `gen_motif_scenicplus`\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "06aee308-54e6-4257-a67a-d73fc9840e0b",
   "metadata": {},
   "source": [
    "3) `rnk`: regions_vs_motifs.rankings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "a2df0359-4a16-4827-88d5-8ead4240ca8e",
   "metadata": {},
   "outputs": [],
   "source": [
    "path = f\"{dir}/dbs/hg38/gen/motif/scenicplus/human_motif_SCREEN.regions_vs_motifs.rankings.feather\"\n",
    "\n",
    "source = pa.memory_map(path, 'r')\n",
    "reader = pa.ipc.open_file(source)\n",
    "schema = reader.schema"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "19677af5-73f6-4f7f-af31-a5b3207e7e37",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First 5 cols: ['chr10:100000176-100000504', 'chr10:100001759-100001930', 'chr10:100004841-100005148', 'chr10:100005876-100006219', 'chr10:100006302-100006644']\n",
      "Last 5 cols: ['chrY:9924284-9924623', 'chrY:9954840-9955041', 'chrY:9959132-9959359', 'chrY:9997981-9998328', 'motifs']\n",
      "Motifs col type: string\n",
      "Metadata: None\n"
     ]
    }
   ],
   "source": [
    "print(\"First 5 cols:\", schema.names[:5])\n",
    "print(\"Last 5 cols:\", schema.names[-5:])\n",
    "print(\"Motifs col type:\", schema.field('motifs').type)\n",
    "print(\"Metadata:\", schema.metadata)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e80de54-ffc4-43ed-9e4c-615dc8cc6021",
   "metadata": {},
   "source": [
    "4) `man` : motif annotations\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "c841cf35-2733-4f2d-a5e9-45f38fc6bc51",
   "metadata": {},
   "outputs": [],
   "source": [
    "man_sample = pd.read_csv(\n",
    "    f\"{dir}/dbs/hg38/gen/motif/scenicplus/motifs-v10nr_clust/nr.hgnc-m0.001-o0.0.tbl\",\n",
    "    sep=\"\\t\",\n",
    "    nrows=10000 \n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "fe7f1234-3c27-4599-8da1-08b7ee8919fc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#motif_id</th>\n",
       "      <th>motif_name</th>\n",
       "      <th>motif_description</th>\n",
       "      <th>source_name</th>\n",
       "      <th>source_version</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>motif_similarity_qvalue</th>\n",
       "      <th>similar_motif_id</th>\n",
       "      <th>similar_motif_description</th>\n",
       "      <th>orthologous_identity</th>\n",
       "      <th>orthologous_gene_name</th>\n",
       "      <th>orthologous_species</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>metacluster_196.3</td>\n",
       "      <td>EcR_usp</td>\n",
       "      <td>EcR/usp</td>\n",
       "      <td>bergman</td>\n",
       "      <td>1.1</td>\n",
       "      <td>HNF4A</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.270042</td>\n",
       "      <td>FBgn0003964</td>\n",
       "      <td>D. melanogaster</td>\n",
       "      <td>gene is orthologous to FBgn0003964 in D. melan...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>metacluster_196.3</td>\n",
       "      <td>EcR_usp</td>\n",
       "      <td>EcR/usp</td>\n",
       "      <td>bergman</td>\n",
       "      <td>1.1</td>\n",
       "      <td>HNF4G</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.276923</td>\n",
       "      <td>FBgn0003964</td>\n",
       "      <td>D. melanogaster</td>\n",
       "      <td>gene is orthologous to FBgn0003964 in D. melan...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>metacluster_196.3</td>\n",
       "      <td>EcR_usp</td>\n",
       "      <td>EcR/usp</td>\n",
       "      <td>bergman</td>\n",
       "      <td>1.1</td>\n",
       "      <td>NR1D1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.157980</td>\n",
       "      <td>FBgn0000546</td>\n",
       "      <td>D. melanogaster</td>\n",
       "      <td>gene is orthologous to FBgn0000546 in D. melan...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           #motif_id motif_name motif_description source_name  source_version  \\\n",
       "0  metacluster_196.3    EcR_usp           EcR/usp     bergman             1.1   \n",
       "1  metacluster_196.3    EcR_usp           EcR/usp     bergman             1.1   \n",
       "2  metacluster_196.3    EcR_usp           EcR/usp     bergman             1.1   \n",
       "\n",
       "  gene_name  motif_similarity_qvalue similar_motif_id  \\\n",
       "0     HNF4A                      0.0              NaN   \n",
       "1     HNF4G                      0.0              NaN   \n",
       "2     NR1D1                      0.0              NaN   \n",
       "\n",
       "  similar_motif_description  orthologous_identity orthologous_gene_name  \\\n",
       "0                       NaN              0.270042           FBgn0003964   \n",
       "1                       NaN              0.276923           FBgn0003964   \n",
       "2                       NaN              0.157980           FBgn0000546   \n",
       "\n",
       "  orthologous_species                                        description  \n",
       "0     D. melanogaster  gene is orthologous to FBgn0003964 in D. melan...  \n",
       "1     D. melanogaster  gene is orthologous to FBgn0003964 in D. melan...  \n",
       "2     D. melanogaster  gene is orthologous to FBgn0000546 in D. melan...  "
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "man_sample.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e0909518-abb6-49a3-9396-8cf08f36dca2",
   "metadata": {},
   "source": [
    "5) `scr`: regions_vs_motifs.scores\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "de8df7b8-07ba-496c-aadc-9375668f531c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dtype: float\n"
     ]
    }
   ],
   "source": [
    "path = f\"{dir}/dbs/hg38/gen/motif/scenicplus/human_motif_SCREEN.regions_vs_motifs.scores.feather\"\n",
    "\n",
    "source = pa.memory_map(path, 'r')\n",
    "reader = pa.ipc.open_file(source)\n",
    "batch = reader.get_batch(0)\n",
    "\n",
    "# Check dtype first\n",
    "region_col = reader.schema.names[0]\n",
    "print(f\"Dtype: {reader.schema.field(region_col).type}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "66867158-e0fa-42c3-abe2-f3b91aa29b8d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First 5 cols: ['chr10:100000176-100000504', 'chr10:100001759-100001930', 'chr10:100004841-100005148', 'chr10:100005876-100006219', 'chr10:100006302-100006644']\n",
      "Last 5 cols: ['chrY:9924284-9924623', 'chrY:9954840-9955041', 'chrY:9959132-9959359', 'chrY:9997981-9998328', 'motifs']\n",
      "Motifs col type: string\n",
      "Metadata: None\n"
     ]
    }
   ],
   "source": [
    "print(\"First 5 cols:\", schema.names[:5])\n",
    "print(\"Last 5 cols:\", schema.names[-5:])\n",
    "print(\"Motifs col type:\", schema.field('motifs').type)\n",
    "print(\"Metadata:\", schema.metadata)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ea32be57-1764-46cf-8239-de2be84bc924",
   "metadata": {},
   "source": [
    "#### 5.2 From `gen_genome_scenicplus`"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "71c3079a-0d23-4147-af47-661bf322f23b",
   "metadata": {},
   "source": [
    "6) `ann`: annotation.tsv\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "676b5536-eee6-4ec1-92d8-8ad8a1413f2c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Transcription_Start_Site</th>\n",
       "      <th>Transcript_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrM</td>\n",
       "      <td>3307</td>\n",
       "      <td>4262</td>\n",
       "      <td>+</td>\n",
       "      <td>MT-ND1</td>\n",
       "      <td>3307</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrM</td>\n",
       "      <td>4470</td>\n",
       "      <td>5511</td>\n",
       "      <td>+</td>\n",
       "      <td>MT-ND2</td>\n",
       "      <td>4470</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrM</td>\n",
       "      <td>5904</td>\n",
       "      <td>7445</td>\n",
       "      <td>+</td>\n",
       "      <td>MT-CO1</td>\n",
       "      <td>5904</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome  Start   End Strand    Gene  Transcription_Start_Site  \\\n",
       "0       chrM   3307  4262      +  MT-ND1                      3307   \n",
       "1       chrM   4470  5511      +  MT-ND2                      4470   \n",
       "2       chrM   5904  7445      +  MT-CO1                      5904   \n",
       "\n",
       "  Transcript_type  \n",
       "0  protein_coding  \n",
       "1  protein_coding  \n",
       "2  protein_coding  "
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ann = pd.read_csv(f\"{dir}/dbs/hg38/gen/genome/scenicplus/annotation.tsv\", sep=\"\\t\")\n",
    "ann.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2c0e816c-ab46-4dfe-bc33-6daa7d2a0975",
   "metadata": {},
   "source": [
    "7) `csz`: chromsizes.tsv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ed38fd3-eb92-4e3c-bb0a-18eca4d41e0e",
   "metadata": {},
   "outputs": [],
   "source": [
    "csz = pd.read_csv(f\"{dir}/dbs/hg38/gen/genome/scenicplus/chromsizes.tsv\", sep=\"\\t\")\n",
    "csz.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "80858241-fabf-4fe9-aea9-b42fd5d32e90",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th># Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Transcript_type</th>\n",
       "      <th>Ensembl_gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrM</td>\n",
       "      <td>3306</td>\n",
       "      <td>3307</td>\n",
       "      <td>MT-ND1</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000198888</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrM</td>\n",
       "      <td>4469</td>\n",
       "      <td>4470</td>\n",
       "      <td>MT-ND2</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000198763</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrM</td>\n",
       "      <td>5903</td>\n",
       "      <td>5904</td>\n",
       "      <td>MT-CO1</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSG00000198804</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  # Chromosome  Start   End    Gene Score Strand Transcript_type  \\\n",
       "0         chrM   3306  3307  MT-ND1     .      +  protein_coding   \n",
       "1         chrM   4469  4470  MT-ND2     .      +  protein_coding   \n",
       "2         chrM   5903  5904  MT-CO1     .      +  protein_coding   \n",
       "\n",
       "   Ensembl_gene_id  \n",
       "0  ENSG00000198888  \n",
       "1  ENSG00000198763  \n",
       "2  ENSG00000198804  "
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tss = pd.read_csv(f\"{dir}/dbs/hg38/gen/genome/scenicplus/tss.tsv\", sep=\"\\t\")\n",
    "tss.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e84857b2-8846-4699-98d3-76a6863e8917",
   "metadata": {},
   "source": [
    "#### 5.3 `extract_case`\n",
    "1) `mdata.h5mu` (multiomics data) will be extracted from  `annotated.h5mu` (rule `annotate_pbmc10k` in Layer 4) <br>\n",
    "Note : `annotated.h5mu` is just grouping scATAC and scRNA together, `mdata.h5mu` is preprocessing + dimensionality reduction (Normalizes RNA and ATAC, filter hvg, Computes UMAP) this is the analysis ready file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "49932a09-71d6-4b0b-b47e-9151372c406a",
   "metadata": {},
   "outputs": [],
   "source": [
    "mdata = mu.read(f\"{dir}/dts/hg38/pbmc10k/cases/all/mdata.h5mu\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "70ae58e5-a382-4ba8-81b6-5353141261eb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre>MuData object with n_obs × n_vars = 9631 × 81920\n",
       "  obs:\t&#x27;batch&#x27;, &#x27;celltype&#x27;\n",
       "  uns:\t&#x27;neighbors&#x27;, &#x27;umap&#x27;\n",
       "  obsm:\t&#x27;X_spectral&#x27;, &#x27;X_umap&#x27;\n",
       "  2 modalities\n",
       "    rna:\t9631 x 16384\n",
       "      layers:\t&#x27;counts&#x27;\n",
       "    atac:\t9631 x 65536\n",
       "      layers:\t&#x27;counts&#x27;</pre>"
      ],
      "text/plain": [
       "MuData object with n_obs × n_vars = 9631 × 81920\n",
       "  obs:\t'batch', 'celltype'\n",
       "  uns:\t'neighbors', 'umap'\n",
       "  obsm:\t'X_spectral', 'X_umap'\n",
       "  2 modalities\n",
       "    rna:\t9631 x 16384\n",
       "      layers:\t'counts'\n",
       "    atac:\t9631 x 65536\n",
       "      layers:\t'counts'"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mdata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "0759c287-e253-48d5-9870-1b6948933f26",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 9631 × 16384\n",
       "    layers: 'counts'"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mdata['rna']     # same nb of cells than annotated but less genes "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "b44a3257-241b-4573-8a90-6e02f45085e5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 9631 × 65536\n",
       "    layers: 'counts'"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mdata['atac']    #  same nb of cells than annotated but less peaks "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7326d31a-b383-41c1-aea5-406e045d3f89",
   "metadata": {},
   "source": [
    "### Layer 6 \n",
    "We now have all the inputs ready, we can run the main script `workflow/scripts/mth/scenicplus/o_mdl.sh` <br>\n",
    "**INPUTS** :\n",
    "- `mdata` : mdata.h5mu (multiomics data) \n",
    "- `blist` : hg38_cre_blacklist.bed\n",
    "- `rnk`: regions_vs_motifs.rankings\n",
    "- `man` : motif annotations\n",
    "- `scr`: regions_vs_motifs.scores\n",
    "- `ann`: annotation.tsv\n",
    "- `csz`: chromsizes.tsv"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a7a02c8c-027d-48f3-a951-ad973cbc4c09",
   "metadata": {},
   "source": [
    "## 3. Script execution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "63b46e02-65fc-404e-a03d-457f9312be90",
   "metadata": {},
   "outputs": [],
   "source": [
    "outdir = f\"{dir}/dts/hg38/pbmc10k/cases/all/runs/scenicplus/\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "58e55807-7785-4479-99fc-2d77b3da5568",
   "metadata": {},
   "source": [
    "### 3.1 pre (preprocessing)\n",
    "\n",
    "Outputs : \n",
    "- mdata.h5mu — processed multiome data (SCENIC+ ready)\n",
    "- rna.h5ad — RNA modality\n",
    "- bool_mat.mtx — binary peak matrix\n",
    "- cres.txt — peak names\n",
    "- brcs.txt — cell barcodes\n",
    "- cistopic_obj.pkl — cisTopic object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "15145ea0-2c25-4779-84a1-3692fc1af565",
   "metadata": {},
   "outputs": [],
   "source": [
    "mdata = mu.read(f\"{outdir}/mdata.h5mu\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "ef402707-5155-4c61-aced-642a9b702121",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre>MuData object with n_obs × n_vars = 9631 × 81540\n",
       "  2 modalities\n",
       "    scRNA:\t9631 x 16384\n",
       "    scATAC:\t9631 x 65156\n",
       "      obs:\t&#x27;cisTopic_nr_frag&#x27;, &#x27;cisTopic_log_nr_frag&#x27;, &#x27;cisTopic_nr_acc&#x27;, &#x27;cisTopic_log_nr_acc&#x27;, &#x27;sample_id&#x27;\n",
       "      var:\t&#x27;Chromosome&#x27;, &#x27;Start&#x27;, &#x27;End&#x27;, &#x27;Width&#x27;, &#x27;cisTopic_nr_frag&#x27;, &#x27;cisTopic_log_nr_frag&#x27;, &#x27;cisTopic_nr_acc&#x27;, &#x27;cisTopic_log_nr_acc&#x27;</pre>"
      ],
      "text/plain": [
       "MuData object with n_obs × n_vars = 9631 × 81540\n",
       "  2 modalities\n",
       "    scRNA:\t9631 x 16384\n",
       "    scATAC:\t9631 x 65156\n",
       "      obs:\t'cisTopic_nr_frag', 'cisTopic_log_nr_frag', 'cisTopic_nr_acc', 'cisTopic_log_nr_acc', 'sample_id'\n",
       "      var:\t'Chromosome', 'Start', 'End', 'Width', 'cisTopic_nr_frag', 'cisTopic_log_nr_frag', 'cisTopic_nr_acc', 'cisTopic_log_nr_acc'"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mdata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "abad41d6-9096-4966-b7c1-7ed60c63f3ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "rna = sc.read_h5ad(f\"{outdir}/rna.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "9ce64906-9a2a-4308-9d79-9ea10354baa5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 9631 × 16384\n",
       "    layers: 'counts'"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rna"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "db5b697f-f490-4ecc-868b-fa703c51440e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<COOrdinate sparse matrix of dtype 'int64'\n",
       "\twith 27464588 stored elements and shape (65195, 9631)>"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bool_mat = scipy.io.mmread(f\"{outdir}/bool_mat.mtx\")\n",
    "bool_mat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "ab9aafc6-78c8-49aa-9c98-cdc03033cc1e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    chr1:865574-866074\n",
       "1    chr1:898634-899134\n",
       "2    chr1:923609-924109\n",
       "Name: 0, dtype: object"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "peaks = pd.read_csv(f\"{outdir}/cres.txt\", header=None)[0]\n",
    "peaks.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "6daac7e9-18ca-4777-b3f6-cab7094a2fbb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    smpl_AAACAGCCAATCCCTT\n",
       "1    smpl_AAACAGCCAATGCGCT\n",
       "2    smpl_AAACAGCCACCAACCG\n",
       "Name: 0, dtype: object"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "barcodes = pd.read_csv(f\"{outdir}/brcs.txt\", header=None)[0]\n",
    "barcodes.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "cf15e72b-720a-41dc-8341-fa989babbd27",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr1:865574-866074</th>\n",
       "      <th>chr1:898634-899134</th>\n",
       "      <th>chr1:923609-924109</th>\n",
       "      <th>chr1:924648-925148</th>\n",
       "      <th>chr1:925418-925918</th>\n",
       "      <th>chr1:940149-940649</th>\n",
       "      <th>chr1:942293-942793</th>\n",
       "      <th>chr1:1014233-1014733</th>\n",
       "      <th>chr1:1032915-1033415</th>\n",
       "      <th>chr1:1040614-1041114</th>\n",
       "      <th>...</th>\n",
       "      <th>chrX:154802027-154802527</th>\n",
       "      <th>chrX:154827496-154827996</th>\n",
       "      <th>chrX:155048964-155049464</th>\n",
       "      <th>chrX:155071042-155071542</th>\n",
       "      <th>chrX:155086399-155086899</th>\n",
       "      <th>chrX:155188720-155189220</th>\n",
       "      <th>chrX:155193852-155194352</th>\n",
       "      <th>chrX:155262910-155263410</th>\n",
       "      <th>chrX:155264267-155264767</th>\n",
       "      <th>chrX:155632407-155632907</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>smpl_AAACAGCCAATCCCTT</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>smpl_AAACAGCCAATGCGCT</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>smpl_AAACAGCCACCAACCG</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>smpl_AAACAGCCAGGATAAC</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>smpl_AAACAGCCAGTTTACG</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>smpl_TTTGTTGGTGACATGC</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>smpl_TTTGTTGGTGTTAAAC</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>smpl_TTTGTTGGTTAGGATT</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>smpl_TTTGTTGGTTGGTTAG</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>smpl_TTTGTTGGTTTGCAGA</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9631 rows × 65195 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "0                      chr1:865574-866074  chr1:898634-899134  \\\n",
       "0                                                               \n",
       "smpl_AAACAGCCAATCCCTT                   0                   0   \n",
       "smpl_AAACAGCCAATGCGCT                   0                   0   \n",
       "smpl_AAACAGCCACCAACCG                   0                   0   \n",
       "smpl_AAACAGCCAGGATAAC                   0                   0   \n",
       "smpl_AAACAGCCAGTTTACG                   0                   0   \n",
       "...                                   ...                 ...   \n",
       "smpl_TTTGTTGGTGACATGC                   0                   0   \n",
       "smpl_TTTGTTGGTGTTAAAC                   0                   0   \n",
       "smpl_TTTGTTGGTTAGGATT                   0                   0   \n",
       "smpl_TTTGTTGGTTGGTTAG                   0                   0   \n",
       "smpl_TTTGTTGGTTTGCAGA                   0                   0   \n",
       "\n",
       "0                      chr1:923609-924109  chr1:924648-925148  \\\n",
       "0                                                               \n",
       "smpl_AAACAGCCAATCCCTT                   0                   0   \n",
       "smpl_AAACAGCCAATGCGCT                   1                   0   \n",
       "smpl_AAACAGCCACCAACCG                   0                   0   \n",
       "smpl_AAACAGCCAGGATAAC                   0                   0   \n",
       "smpl_AAACAGCCAGTTTACG                   0                   0   \n",
       "...                                   ...                 ...   \n",
       "smpl_TTTGTTGGTGACATGC                   0                   0   \n",
       "smpl_TTTGTTGGTGTTAAAC                   0                   1   \n",
       "smpl_TTTGTTGGTTAGGATT                   0                   0   \n",
       "smpl_TTTGTTGGTTGGTTAG                   0                   0   \n",
       "smpl_TTTGTTGGTTTGCAGA                   0                   0   \n",
       "\n",
       "0                      chr1:925418-925918  chr1:940149-940649  \\\n",
       "0                                                               \n",
       "smpl_AAACAGCCAATCCCTT                   0                   1   \n",
       "smpl_AAACAGCCAATGCGCT                   0                   0   \n",
       "smpl_AAACAGCCACCAACCG                   0                   0   \n",
       "smpl_AAACAGCCAGGATAAC                   0                   0   \n",
       "smpl_AAACAGCCAGTTTACG                   0                   0   \n",
       "...                                   ...                 ...   \n",
       "smpl_TTTGTTGGTGACATGC                   0                   0   \n",
       "smpl_TTTGTTGGTGTTAAAC                   0                   0   \n",
       "smpl_TTTGTTGGTTAGGATT                   0                   0   \n",
       "smpl_TTTGTTGGTTGGTTAG                   0                   0   \n",
       "smpl_TTTGTTGGTTTGCAGA                   0                   0   \n",
       "\n",
       "0                      chr1:942293-942793  chr1:1014233-1014733  \\\n",
       "0                                                                 \n",
       "smpl_AAACAGCCAATCCCTT                   0                     0   \n",
       "smpl_AAACAGCCAATGCGCT                   0                     0   \n",
       "smpl_AAACAGCCACCAACCG                   0                     0   \n",
       "smpl_AAACAGCCAGGATAAC                   0                     0   \n",
       "smpl_AAACAGCCAGTTTACG                   0                     0   \n",
       "...                                   ...                   ...   \n",
       "smpl_TTTGTTGGTGACATGC                   0                     0   \n",
       "smpl_TTTGTTGGTGTTAAAC                   0                     0   \n",
       "smpl_TTTGTTGGTTAGGATT                   0                     0   \n",
       "smpl_TTTGTTGGTTGGTTAG                   0                     0   \n",
       "smpl_TTTGTTGGTTTGCAGA                   0                     0   \n",
       "\n",
       "0                      chr1:1032915-1033415  chr1:1040614-1041114  ...  \\\n",
       "0                                                                  ...   \n",
       "smpl_AAACAGCCAATCCCTT                     0                     0  ...   \n",
       "smpl_AAACAGCCAATGCGCT                     0                     0  ...   \n",
       "smpl_AAACAGCCACCAACCG                     0                     0  ...   \n",
       "smpl_AAACAGCCAGGATAAC                     1                     0  ...   \n",
       "smpl_AAACAGCCAGTTTACG                     0                     0  ...   \n",
       "...                                     ...                   ...  ...   \n",
       "smpl_TTTGTTGGTGACATGC                     0                     0  ...   \n",
       "smpl_TTTGTTGGTGTTAAAC                     0                     0  ...   \n",
       "smpl_TTTGTTGGTTAGGATT                     0                     0  ...   \n",
       "smpl_TTTGTTGGTTGGTTAG                     0                     0  ...   \n",
       "smpl_TTTGTTGGTTTGCAGA                     0                     0  ...   \n",
       "\n",
       "0                      chrX:154802027-154802527  chrX:154827496-154827996  \\\n",
       "0                                                                           \n",
       "smpl_AAACAGCCAATCCCTT                         0                         0   \n",
       "smpl_AAACAGCCAATGCGCT                         0                         0   \n",
       "smpl_AAACAGCCACCAACCG                         0                         0   \n",
       "smpl_AAACAGCCAGGATAAC                         0                         0   \n",
       "smpl_AAACAGCCAGTTTACG                         0                         0   \n",
       "...                                         ...                       ...   \n",
       "smpl_TTTGTTGGTGACATGC                         0                         0   \n",
       "smpl_TTTGTTGGTGTTAAAC                         0                         0   \n",
       "smpl_TTTGTTGGTTAGGATT                         1                         0   \n",
       "smpl_TTTGTTGGTTGGTTAG                         0                         0   \n",
       "smpl_TTTGTTGGTTTGCAGA                         0                         0   \n",
       "\n",
       "0                      chrX:155048964-155049464  chrX:155071042-155071542  \\\n",
       "0                                                                           \n",
       "smpl_AAACAGCCAATCCCTT                         0                         1   \n",
       "smpl_AAACAGCCAATGCGCT                         0                         0   \n",
       "smpl_AAACAGCCACCAACCG                         0                         0   \n",
       "smpl_AAACAGCCAGGATAAC                         0                         0   \n",
       "smpl_AAACAGCCAGTTTACG                         0                         0   \n",
       "...                                         ...                       ...   \n",
       "smpl_TTTGTTGGTGACATGC                         0                         1   \n",
       "smpl_TTTGTTGGTGTTAAAC                         0                         0   \n",
       "smpl_TTTGTTGGTTAGGATT                         0                         1   \n",
       "smpl_TTTGTTGGTTGGTTAG                         0                         0   \n",
       "smpl_TTTGTTGGTTTGCAGA                         0                         1   \n",
       "\n",
       "0                      chrX:155086399-155086899  chrX:155188720-155189220  \\\n",
       "0                                                                           \n",
       "smpl_AAACAGCCAATCCCTT                         0                         0   \n",
       "smpl_AAACAGCCAATGCGCT                         0                         0   \n",
       "smpl_AAACAGCCACCAACCG                         0                         0   \n",
       "smpl_AAACAGCCAGGATAAC                         0                         0   \n",
       "smpl_AAACAGCCAGTTTACG                         0                         0   \n",
       "...                                         ...                       ...   \n",
       "smpl_TTTGTTGGTGACATGC                         0                         0   \n",
       "smpl_TTTGTTGGTGTTAAAC                         0                         0   \n",
       "smpl_TTTGTTGGTTAGGATT                         0                         0   \n",
       "smpl_TTTGTTGGTTGGTTAG                         0                         0   \n",
       "smpl_TTTGTTGGTTTGCAGA                         0                         0   \n",
       "\n",
       "0                      chrX:155193852-155194352  chrX:155262910-155263410  \\\n",
       "0                                                                           \n",
       "smpl_AAACAGCCAATCCCTT                         0                         0   \n",
       "smpl_AAACAGCCAATGCGCT                         0                         0   \n",
       "smpl_AAACAGCCACCAACCG                         0                         0   \n",
       "smpl_AAACAGCCAGGATAAC                         0                         0   \n",
       "smpl_AAACAGCCAGTTTACG                         0                         0   \n",
       "...                                         ...                       ...   \n",
       "smpl_TTTGTTGGTGACATGC                         0                         0   \n",
       "smpl_TTTGTTGGTGTTAAAC                         0                         0   \n",
       "smpl_TTTGTTGGTTAGGATT                         0                         0   \n",
       "smpl_TTTGTTGGTTGGTTAG                         0                         0   \n",
       "smpl_TTTGTTGGTTTGCAGA                         0                         0   \n",
       "\n",
       "0                      chrX:155264267-155264767  chrX:155632407-155632907  \n",
       "0                                                                          \n",
       "smpl_AAACAGCCAATCCCTT                         0                         0  \n",
       "smpl_AAACAGCCAATGCGCT                         0                         0  \n",
       "smpl_AAACAGCCACCAACCG                         0                         0  \n",
       "smpl_AAACAGCCAGGATAAC                         0                         0  \n",
       "smpl_AAACAGCCAGTTTACG                         0                         0  \n",
       "...                                         ...                       ...  \n",
       "smpl_TTTGTTGGTGACATGC                         1                         0  \n",
       "smpl_TTTGTTGGTGTTAAAC                         0                         0  \n",
       "smpl_TTTGTTGGTTAGGATT                         0                         0  \n",
       "smpl_TTTGTTGGTTGGTTAG                         1                         0  \n",
       "smpl_TTTGTTGGTTTGCAGA                         0                         0  \n",
       "\n",
       "[9631 rows x 65195 columns]"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bool_df = pd.DataFrame(bool_mat.toarray().T, index=barcodes, columns=peaks)\n",
    "bool_df    # cells x peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "549fd818-f10a-409e-baba-501c13594fbe",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "\n",
    "with open(f\"{outdir}/cistopic_obj.pkl\", \"rb\") as f:\n",
    "    cistopic_obj = pickle.load(f)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8d46ccb6-fdfd-4fe2-be68-887d9ed697fd",
   "metadata": {},
   "source": [
    "### p2g - CRE (peak) to gene\n",
    "Outputs :\n",
    "\n",
    "- space.tsv (search space file, links peaks to genes)\n",
    "- rg_adj.tsv (region gene adj)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "5cedf217-012e-4842-ae22-193c4d996c53",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Distance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1:94237395-94237895</td>\n",
       "      <td>ABCD3</td>\n",
       "      <td>[180744]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1:94384797-94385297</td>\n",
       "      <td>ABCD3</td>\n",
       "      <td>[33342]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1:94418054-94418554</td>\n",
       "      <td>ABCD3</td>\n",
       "      <td>[85]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Name   Gene  Distance\n",
       "0  chr1:94237395-94237895  ABCD3  [180744]\n",
       "1  chr1:94384797-94385297  ABCD3   [33342]\n",
       "2  chr1:94418054-94418554  ABCD3      [85]"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "space = pd.read_csv(f\"{outdir}/space.tsv\", sep=\"\\t\")\n",
    "space.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "3e71a7bd-0c66-485c-ab26-6e1551c1c975",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>target</th>\n",
       "      <th>region</th>\n",
       "      <th>importance</th>\n",
       "      <th>rho</th>\n",
       "      <th>importance_x_rho</th>\n",
       "      <th>importance_x_abs_rho</th>\n",
       "      <th>Distance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A1BG</td>\n",
       "      <td>chr19:58513856-58514356</td>\n",
       "      <td>0.119557</td>\n",
       "      <td>0.023117</td>\n",
       "      <td>0.002764</td>\n",
       "      <td>0.002764</td>\n",
       "      <td>[160614]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A1BG</td>\n",
       "      <td>chr19:58347446-58347946</td>\n",
       "      <td>0.251514</td>\n",
       "      <td>0.086522</td>\n",
       "      <td>0.021762</td>\n",
       "      <td>0.021762</td>\n",
       "      <td>[0]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A1BG</td>\n",
       "      <td>chr19:58229188-58229688</td>\n",
       "      <td>0.076964</td>\n",
       "      <td>0.012665</td>\n",
       "      <td>0.000975</td>\n",
       "      <td>0.000975</td>\n",
       "      <td>[-115740]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  target                   region  importance       rho  importance_x_rho  \\\n",
       "0   A1BG  chr19:58513856-58514356    0.119557  0.023117          0.002764   \n",
       "1   A1BG  chr19:58347446-58347946    0.251514  0.086522          0.021762   \n",
       "2   A1BG  chr19:58229188-58229688    0.076964  0.012665          0.000975   \n",
       "\n",
       "   importance_x_abs_rho   Distance  \n",
       "0              0.002764   [160614]  \n",
       "1              0.021762        [0]  \n",
       "2              0.000975  [-115740]  "
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rg_adj = pd.read_csv(f\"{outdir}/rg_adj.tsv\",sep=\"\\t\")\n",
    "rg_adj.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "05b71263-d754-4aee-8941-b4e47b72ca43",
   "metadata": {},
   "source": [
    "### 3.3 tfb - TF to CRE binding prediction\n",
    "Outputs :\n",
    "\n",
    "- topic_*  (raw topic files)\n",
    "- otsu/    (binarized per-topic region sets)\n",
    "- cistarget.hdf5  (Topic × TF enrichment scores)\n",
    "<br><br>\n",
    "- `tfs.txt` :  List of TFs that passed motif enrichment thresholds\n",
    "- `direct.h5ad` : Annotated dataset with direct TF → peak associations\n",
    "- `extended.h5ad` :  Annotated dataset with extended TF → peak associations (orthology or additional motifs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "af085a94-b73c-4f7c-b5b0-d4091df995dd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<HDF5 file \"cistarget.hdf5\" (mode r)>"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ct = h5py.File(f\"{outdir}/cistarget.hdf5\", \"r\")\n",
    "ct"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "b0668d1f-f8d9-450a-8e39-be1b240ffe2e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['otsu_region_Topic_10_binarized',\n",
       " 'otsu_region_Topic_11_binarized',\n",
       " 'otsu_region_Topic_12_binarized',\n",
       " 'otsu_region_Topic_13_binarized',\n",
       " 'otsu_region_Topic_14_binarized']"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(ct.keys())[:5] # 5 first topics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "8d034ea2-7cfd-423f-a2eb-4f41c5d47c13",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<HDF5 group \"/otsu_region_Topic_10_binarized\" (5 members)>"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "topic10 = ct[\"otsu_region_Topic_10_binarized\"]\n",
    "topic10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "4da6c3d4-7b8b-436d-bae2-76bb085cb42e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['cistromes', 'metadata', 'motif_enrichment', 'motif_hits', 'regions_to_db']"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(topic10.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "5f59be8e-3061-4824-aee6-a9ea22f9e324",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TF</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BCL11A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FOXO6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CEBPD</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       TF\n",
       "0  BCL11A\n",
       "1   FOXO6\n",
       "2   CEBPD"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tfs = pd.read_csv(f\"{outdir}/tfs.txt\", header=None, names=[\"TF\"])\n",
    "tfs.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "20bfe318-555a-49d6-b848-50b473b3dfb5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 36806 × 630\n",
       "    var: 'motifs', 'is_extended'"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "direct_adata = sc.read_h5ad(f\"{outdir}/direct.h5ad\")\n",
    "direct_adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "1f87e53b-d776-4088-8e6c-14e208006c36",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr22:43004233-43004733</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr3:118851741-118852241</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr11:64851176-64851676</th>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: [chr22:43004233-43004733, chr3:118851741-118852241, chr11:64851176-64851676]"
      ]
     },
     "execution_count": 105,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "direct_adata.obs.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "195cea32-5c53-4dec-80da-881c28c6aa68",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>motifs</th>\n",
       "      <th>is_extended</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ACAA1</th>\n",
       "      <td>metacluster_67.2</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AHCTF1</th>\n",
       "      <td>cisbp__M00750</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AIRE</th>\n",
       "      <td>metacluster_129.1</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   motifs  is_extended\n",
       "ACAA1    metacluster_67.2        False\n",
       "AHCTF1      cisbp__M00750        False\n",
       "AIRE    metacluster_129.1        False"
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "direct_adata.var.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "id": "cfaa8461-ea53-468c-ad34-b6bc6854c1cc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Compressed Sparse Column sparse matrix of dtype 'bool'\n",
       "\twith 1852230 stored elements and shape (36806, 630)>"
      ]
     },
     "execution_count": 110,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "direct_adata.X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "id": "006a02de-7254-4f0a-bcf1-7be85a6d5d92",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 32987 × 275\n",
       "    var: 'motifs', 'is_extended'"
      ]
     },
     "execution_count": 107,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "extended_adata = sc.read_h5ad(f\"{outdir}/extended.h5ad\")\n",
    "extended_adata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "37dab94a-1902-410c-891b-cd3e7c3f62d3",
   "metadata": {},
   "source": [
    "### 3.4 mdl - Modeling (final TF-Gene links)\n",
    "Outputs :\n",
    "\n",
    "- `tg_adg` : TF-TG-Score matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "id": "27035f43-044c-43ae-8875-739e9f66b9e6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TF</th>\n",
       "      <th>target</th>\n",
       "      <th>importance</th>\n",
       "      <th>regulation</th>\n",
       "      <th>rho</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACAA1</td>\n",
       "      <td>BID</td>\n",
       "      <td>0.091457</td>\n",
       "      <td>1</td>\n",
       "      <td>0.086696</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACAA1</td>\n",
       "      <td>ST14</td>\n",
       "      <td>0.103735</td>\n",
       "      <td>1</td>\n",
       "      <td>0.044573</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACAA1</td>\n",
       "      <td>NOP58</td>\n",
       "      <td>0.034211</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.016389</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      TF target  importance  regulation       rho\n",
       "0  ACAA1    BID    0.091457           1  0.086696\n",
       "1  ACAA1   ST14    0.103735           1  0.044573\n",
       "2  ACAA1  NOP58    0.034211           0 -0.016389"
      ]
     },
     "execution_count": 112,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tg_adj = pd.read_csv(f\"{outdir}/tg_adj.tsv\", sep=\"\\t\")\n",
    "tg_adj.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f0cda07-7750-45a9-a68e-2026c8386c6e",
   "metadata": {},
   "source": [
    "### 3.5 Fine tuned mdl\n",
    "\n",
    "- `egrn.tsv` : final raw TF → gene network\n",
    "- `o_scenicplus.o_scenicplus.o_scenicplus.o_scenicplus.mdl.csv` final GRN, output of the rule ! Basically a filtering of the `egrn.tsv` (sorted by `triplet_rank`)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "id": "163ba97c-82b6-4885-98aa-675077c8d09b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Region</th>\n",
       "      <th>Gene</th>\n",
       "      <th>importance_R2G</th>\n",
       "      <th>rho_R2G</th>\n",
       "      <th>importance_x_rho</th>\n",
       "      <th>importance_x_abs_rho</th>\n",
       "      <th>TF</th>\n",
       "      <th>is_extended</th>\n",
       "      <th>eRegulon_name</th>\n",
       "      <th>Gene_signature_name</th>\n",
       "      <th>Region_signature_name</th>\n",
       "      <th>importance_TF2G</th>\n",
       "      <th>regulation</th>\n",
       "      <th>rho_TF2G</th>\n",
       "      <th>triplet_rank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr16:20691493-20691993</td>\n",
       "      <td>LYRM1</td>\n",
       "      <td>0.153658</td>\n",
       "      <td>0.076111</td>\n",
       "      <td>0.011695</td>\n",
       "      <td>0.011695</td>\n",
       "      <td>ARID3A</td>\n",
       "      <td>False</td>\n",
       "      <td>ARID3A_direct_+/+</td>\n",
       "      <td>ARID3A_direct_+/+_(13g)</td>\n",
       "      <td>ARID3A_direct_+/+_(16r)</td>\n",
       "      <td>0.503523</td>\n",
       "      <td>1</td>\n",
       "      <td>0.054331</td>\n",
       "      <td>15496</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr5:135462204-135462704</td>\n",
       "      <td>TIFAB</td>\n",
       "      <td>0.209777</td>\n",
       "      <td>0.115862</td>\n",
       "      <td>0.024305</td>\n",
       "      <td>0.024305</td>\n",
       "      <td>ARID3A</td>\n",
       "      <td>False</td>\n",
       "      <td>ARID3A_direct_+/+</td>\n",
       "      <td>ARID3A_direct_+/+_(13g)</td>\n",
       "      <td>ARID3A_direct_+/+_(16r)</td>\n",
       "      <td>0.344850</td>\n",
       "      <td>1</td>\n",
       "      <td>0.050536</td>\n",
       "      <td>12059</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr3:130002927-130003427</td>\n",
       "      <td>TMCC1</td>\n",
       "      <td>0.024978</td>\n",
       "      <td>0.062072</td>\n",
       "      <td>0.001550</td>\n",
       "      <td>0.001550</td>\n",
       "      <td>ARID3A</td>\n",
       "      <td>False</td>\n",
       "      <td>ARID3A_direct_+/+</td>\n",
       "      <td>ARID3A_direct_+/+_(13g)</td>\n",
       "      <td>ARID3A_direct_+/+_(16r)</td>\n",
       "      <td>0.339438</td>\n",
       "      <td>1</td>\n",
       "      <td>0.082566</td>\n",
       "      <td>37174</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Region   Gene  importance_R2G   rho_R2G  \\\n",
       "0   chr16:20691493-20691993  LYRM1        0.153658  0.076111   \n",
       "1  chr5:135462204-135462704  TIFAB        0.209777  0.115862   \n",
       "2  chr3:130002927-130003427  TMCC1        0.024978  0.062072   \n",
       "\n",
       "   importance_x_rho  importance_x_abs_rho      TF  is_extended  \\\n",
       "0          0.011695              0.011695  ARID3A        False   \n",
       "1          0.024305              0.024305  ARID3A        False   \n",
       "2          0.001550              0.001550  ARID3A        False   \n",
       "\n",
       "       eRegulon_name      Gene_signature_name    Region_signature_name  \\\n",
       "0  ARID3A_direct_+/+  ARID3A_direct_+/+_(13g)  ARID3A_direct_+/+_(16r)   \n",
       "1  ARID3A_direct_+/+  ARID3A_direct_+/+_(13g)  ARID3A_direct_+/+_(16r)   \n",
       "2  ARID3A_direct_+/+  ARID3A_direct_+/+_(13g)  ARID3A_direct_+/+_(16r)   \n",
       "\n",
       "   importance_TF2G  regulation  rho_TF2G  triplet_rank  \n",
       "0         0.503523           1  0.054331         15496  \n",
       "1         0.344850           1  0.050536         12059  \n",
       "2         0.339438           1  0.082566         37174  "
      ]
     },
     "execution_count": 115,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "egrn = pd.read_csv(f\"{outdir}/egrn.tsv\", sep=\"\\t\")\n",
    "egrn.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "id": "45f87924-599a-48df-b727-c4934f52e5d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "path = f\"{dir}/dts/hg38/pbmc10k/cases/all/runs/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "id": "40073fc5-df64-413c-9101-cac17fdc2c64",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(42219, 5)"
      ]
     },
     "execution_count": 126,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mdl_df = pd.read_csv(f\"{path}/o_scenicplus.o_scenicplus.o_scenicplus.o_scenicplus.mdl.csv\")\n",
    "mdl_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "id": "0995a490-2b7c-4720-8be9-41b6cf289f8b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>cre</th>\n",
       "      <th>target</th>\n",
       "      <th>score</th>\n",
       "      <th>pval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CREB5</td>\n",
       "      <td>chr7-28409015-28409515</td>\n",
       "      <td>CREB5</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ETV6</td>\n",
       "      <td>chr12-82966084-82966584</td>\n",
       "      <td>TMTC2</td>\n",
       "      <td>0.999976</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SPIB</td>\n",
       "      <td>chr2-17804303-17804803</td>\n",
       "      <td>GEN1</td>\n",
       "      <td>0.999953</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SPIB</td>\n",
       "      <td>chr2-164840367-164840867</td>\n",
       "      <td>COBLL1</td>\n",
       "      <td>0.999929</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TCF4</td>\n",
       "      <td>chr1-4796368-4796868</td>\n",
       "      <td>AJAP1</td>\n",
       "      <td>0.999905</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  source                       cre  target     score  pval\n",
       "0  CREB5    chr7-28409015-28409515   CREB5  1.000000  0.01\n",
       "1   ETV6   chr12-82966084-82966584   TMTC2  0.999976  0.01\n",
       "2   SPIB    chr2-17804303-17804803    GEN1  0.999953  0.01\n",
       "3   SPIB  chr2-164840367-164840867  COBLL1  0.999929  0.01\n",
       "4   TCF4      chr1-4796368-4796868   AJAP1  0.999905  0.01"
      ]
     },
     "execution_count": 127,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# source : TF\n",
    "# target : TG\n",
    "mdl_df.head()#.sort_values(by='score')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d12ca398-643f-4fa5-9874-c47e55e7b7ab",
   "metadata": {},
   "source": [
    "This is the expected ouput for GRETA :\n",
    "\n",
    "```bash\n",
    "if [ $? -eq 124 ]; then\n",
    "    awk 'BEGIN {{ print \"source,target,score,pval\" }}' > {output.out}\n",
    "fi``"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b3019ed9-9f61-418f-a010-790dc4bca8a0",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
